apps/zotonic_core/src/support/z_nonce.erl

```diff
@@ -0,0 +1,291 @@
+% they are all emptied. Generation garbage collection by periodically
+% emptying the oldest table.
+-define(NONCE_TABLES, [
+        z_nonce_0, z_nonce_1, z_nonce_2, z_nonce_3, z_nonce_4,
+        z_nonce_5, z_nonce_6, z_nonce_7, z_nonce_8, z_nonce_9
+    ]).
+
+% Number of seconds an nonce is valid. This can vary 10% due to the
+% periodic cleanup of the nonce tables. The 1000 sec, with a variety
+% of 10% gives 15 minutes reuse checks.
+-define(NONCE_TIMEOUT, 1000).
+
+% Default validity of a generated nonce value, must be used within this
+% period otherwise the key will be expired. This period is shorter than
+% the max period we keep registered (used) keys in the generational ets
+% tables so that we can guarantee the unique use of these keys.
+-define(NONCE_VALIDITY, 900).
+
+% Maximum number of nonce tokens allowed in a single generational nonce
+% table. The routines will give an overload error if passed this.
+-define(NONCE_OVERLOAD, 10_000_000).
+
+% Maximum size for random nonce values from other services.
+-define(NONCE_ANY_MAX_SIZE, 50).
+
+
+%%%--------------------------------------------------------------------------
+%%% API
+%%%--------------------------------------------------------------------------
+
+-spec init_nonce_tables() -> ok.
+%% @doc Initialize the nonce tables. Calles by zotonic_core_sup, which process
+%% will be the owner of these tables.
+init_nonce_tables() ->
+    nonce_next_cleanup(),
+    lists:foreach(
+        fun(N) ->
+            ets:new(N, [ named_table, public ])
+        end,
+        ?NONCE_TABLES).
+
+-spec nonce() -> Nonce when
+    Nonce :: binary().
+%% @doc Return a new nonce value, strictly valid for the how long we keep the
+%% used nonce values. This gives maximum protection against replay attacks.
+nonce() ->
+    nonce(?NONCE_VALIDITY).
+
+-spec nonce(Timeout) -> Nonce when
+    Timeout :: integer(),
+    Nonce :: binary().
+%% @doc Return a new nonce value, valid for about the given number of seconds.
+nonce(Timeout) ->
+    Number = nonce_generation(Timeout),
+    Random = z_ids:id(10),
+    Key = <<(integer_to_binary(Number))/binary, "-", Random/binary>>,
+    Hash = crypto:mac(hmac, sha256, Key, nonce_secret()),
+    HashHex = base64url:encode(Hash),
+    <<Key/binary, "-", HashHex/binary>>.
+
+-spec register(Nonce) -> ok | {error, Reason} when
+    Nonce :: binary(),
+    Reason :: duplicate | overload | key | expired.
+%% @doc Register a nonce for use, will be remembered for the next ?NONCE_TIMEOUT
+%% seconds or until unregistered. The nonce must be generated with nonce/1 and
+%% will be checked if it was expired.
+register(<<>>) ->
+    {error, key};
+register(Nonce) ->
+    Oldest = nonce_oldest_generation(),
+    case split_nonce(Nonce) of
+        {Number, Random, Hash} when Number >= Oldest ->
+            CheckKey = <<(integer_to_binary(Number))/binary, "-", Random/binary>>,
+            CheckHash = crypto:mac(hmac, sha256, CheckKey, nonce_secret()),
```

Comment:
If you want to have a cryptographically secure nonce, then you have to make sure you can also check nonces after restarts and crashes. The ets tables will disappear after a crash, but the nonce_secret will survive the restart. This means that the nonce can be used multiple times. A solution for this is to create a new secret for every restart. It has as downside that all in flight nonces are invalidated.

Reply:
I much prefer to not have the inflight nonces be invalidated.
As that would result all kinds of requests to fail after a restart.

We could add an API for a "strong" nonce - where we re-generate the key after each restart.